### ISSUE TITLE
Basic Transportation System

### ISSUE THREAD

**Issue by u04576617** [COS214-Project-2024/WorksOnMyMachineTeam#26]
The simulation must allow for the development and management of transportation systems:
- [x] The user should be able to build roads.
- [x] Public transport should be available in some form and identified by the system to tell the user.
- [ ] Train systems and airports must be placeable (the user should be able to create them somehow).

**Comment by u04576617** [comment]
@Thunder2505 How are you planning on making the user place roads in the GUI? Do you think like a line or something else? Depending on that I will change how the transportation system functions.

**Comment by Thunder2505** [comment]
It is going to be a line, which I was planning on linking to other roads using a linked list and some gui manipulation, to allow snapping to other roads, but I can do it in a different way if you need something to change

**Comment by u04576617** [comment]
> It is going to be a line, which I was planning on linking to other roads using a linked list and some gui manipulation, to allow snapping to other roads, but I can do it in a different way if you need something to change

@Thunder2505  No that could work. I'll add a way to add connected roads, maybe just a vector of road pointers for a pseudo linked list and then go from there. One thing we would need to do then is have long pieces of roads be split up into smaller pieces to make traffic work correctly. Could be a nice spot to add in the road composite.

Here's what I'm thinking:
Say the user want to build a road 220 units long, we split it into sections of for example 50 max, so 50+50+50+50+20. Then we will create a composite road consisting of so many pieces. Of course, this would cause issues when adding connected (say addConnectedRoad is the function name) roads, so maybe we can add a "distance" parameter to the addConnectedRoad(). So then if we needed to add an intersection 130m down the road, we pass 130 and it will then know to place it on section 3. 

For this to work each road would need an additional "start" and "end" position to know where to start from, maybe just 4 x/y ints as that is what the gui wants? Lemme know what you think.

**Comment by Thunder2505** [comment]
So a chain of responsibility?
We can do that, it's fine. I'm not exactly sure how you want to do it though. I assume this would happen in the road's constructor.

**Comment by u04576617** [comment]
Yes exactly. I think each time we create a road, we use RoadComposite. I added a parameter to the constructor for type, whether it is highway, main or residential. Then RoadComposite will handle the rest. Even if the road is short, it is still composite, I think that would simplify things a lot.

**Comment by u04576617** [comment]
@Thunder2505 Is the buildings class going to have coordinates? I need to know where the buildings are to handle navigation. Literally just the x/y coordinates will be more than enough.

**Comment by Thunder2505** [comment]
Navigation?
What are we navigating to?

**Comment by u04576617** [comment]
@Thunder2505 @Amber21457752 I think for trains and airports we can keep things simple by:

- Allowing citizens to move into/leave the city via the airport (so it is tied to city growth)
- Trains just "Teleport" citizens from point A to B, thus avoiding all traffic
- Airports and trains also need to do freight, maybe we can import/export various resources using them? Maybe trains export and planes import? Something that will be easy to implement would probably be best

**Comment by Thunder2505** [comment]
Yeah, that's better than actually implementing transport routs

**Comment by u04576617** [comment]
@Thunder2505  Cool, you'll add a way to create an airport?

Or maybe instead of that our city starts with an airport? It does not really exist anywhere but the logic does?

And for trains I think maybe we can add trainstations as buildings, I can add the logic to the existing transport system. I just need to know where they are placed.

**Comment by u04576617** [comment]
@Thunder2505 I have added a building for a Trainstation. Usage is very simple, you only need to create the building passing it's range of effect and set it's coordinates, everything else is done automatically.
